src/process_unix.cpp

```diff
@@ -0,0 +1,550 @@
+            }
+            if (!startup._stdout.redirected()) {
+                close_fd(pstdout[PIPE_READ]);
+            }
+
+            dup2(pstdin[PIPE_READ], STDIN_FILENO);
+            dup2(pstdout[PIPE_WRITE], STDOUT_FILENO);
+
+            /*
+             * pay special attention to stderr,
+             * there are 2 cases:
+             *      1. redirect stderr to stdout
+             *      2. redirect stderr to a file
+             */
+            if (startup.merge_outputs) {
+                // redirect stderr to stdout
+                dup2(pstdout[PIPE_WRITE], STDERR_FILENO);
+            } else {
+                // redirect stderr to a file
+                if (!startup._stderr.redirected()) {
+                    close_fd(pstderr[PIPE_READ]);
+                }
+                dup2(pstderr[PIPE_WRITE], STDERR_FILENO);
+            }
+
+            close_fd(pstdin[PIPE_READ]);
+            close_fd(pstdout[PIPE_WRITE]);
+            close_fd(pstderr[PIPE_WRITE]);
+
+            // command-line and environments
+            size_t asize = startup._cmdline.size();
+            size_t esize = startup._env.size();
+            char *argv[asize + 1];
+            char *envp[esize + 1];
+
+            // argv and envp are always terminated with a nullptr
+            argv[asize] = nullptr;
+            envp[esize] = nullptr;
+
+            // copy command-line arguments
+            for (std::size_t i = 0; i < asize; ++i) {
+                argv[i] = const_cast<char *>(startup._cmdline[i].c_str());
+            }
+
+            // copy environment variables
+            std::vector<std::string> envs;
+            std::stringstream buffer;
+            for (const auto &e : startup._env) {
+                buffer.str("");
+                buffer.clear();
+                buffer << e.first << "=" << e.second;
+                envs.emplace_back(buffer.str());
+            }
+
+            for (std::size_t i = 0; i < esize; ++i) {
+                envp[i] = const_cast<char *>(envs[i].c_str());
+            }
+
+            // close everything
+            if (!close_all_descriptors(STDERR_FILENO + 1, fail_fd)) {
+                // try luck failed, close the old way
+                int max_fd = static_cast<int>(sysconf(_SC_OPEN_MAX));
+                for (int fd = STDERR_FILENO + 1; fd < max_fd; fd++) {
+                    // do not close fail pipe
+                    if (fd == fail_fd) {
+                        continue;
+                    }
+                    if (close(fd) == -1 && errno != EBADF) {
+                        // oops, we cannot close this fd
+                        // TODO: should we report this as an error?
+                        continue;
+                    }
+                }
+            }
+
+            // change cwd
+            if (chdir(startup._cwd.c_str()) != 0) {
+                // do not use mpp::throw_ex
+                throw mpp::runtime_error();
```

Comment:
Consider using `mpp::throw_ex` because it can handle noexcept situations.

Reply:
No. noexcept will be handle with `std::terminate`.
And I don't know what happens inside this mysterious C++ function.
Here the parent are waiting the child to perform some IO operations (e.g. close or write) on the fail pipe, BUT I am not sure if **ALL** `std::terminate` will do the cleanup job.

I using raw throw here is just to jump the control flow to the cleanup aera, which will ensure
the parent-child-communication pipe will be closed properly, thus ensuring the parent process will not get stuck in `read_fully`.

Reply:
So the best way might be `goto`...?

Reply:
Fuck me, solved.